arduino/wrappers/Utils.h

```diff
@@ -0,0 +1,370 @@
+*/
+
+#pragma once
+
+// This header is part of the function objects library and provides the standard hash function.
+#include <functional>
+
+// In case of Arduino, use Arduino syntax.
+#if defined(ARDUINO)
+#include <Arduino.h>
+#else
+
+// else use the syntax for C++
+#include <string>
+using String = std::string;
+#endif
+
+// Global variable declarations.
+const auto ONE_SECOND = 1000UL;
+const int NO_OBSTACLE_VALUE = 0;
+const int FRONT_STOP_DISTANCE = 70;
+const int BACK_STOP_DISTANCE = 50;
+const float STOPPING_SPEED = 0.3;
+const int SIDE_REACT_DISTANCE = 35;
+const float MAX_SPEED = 1.845;
+float initialSpeed = 1.845;
+
+// Wrapper for ArduinoRunTime
+struct ArduinoRunWrapper {
+    virtual ~ArduinoRunWrapper() = default;
+
+    virtual long millis()                = 0;
+    virtual void delay(int n)            = 0;
+};
+
+// Wrapper for UltraSoundSensor
+struct UltraSoundWrapper {
+
+    virtual ~UltraSoundWrapper() = default;
+
+    virtual int getDistance()          = 0;
+
+};
+
+//Wrapper for InfraredSensor
+struct InfraredSensorWrapper {
+
+    virtual ~InfraredSensorWrapper() = default;
+
+    virtual int getDistance()          = 0;
+
+};
+
+// Wrapper for ArduinoSerial
+struct SerialWrapper {
+
+    virtual ~SerialWrapper() = default;
+
+    virtual void println(String message) = 0;
+    virtual void begin(int n) = 0;
+    virtual bool available() = 0;
+    virtual char readStringUntil(char singleQuotion) = 0;
+};
+
+//Wrapper for MQTT
+struct MqttWrapper {
+
+  virtual ~MqttWrapper() = default;
+
+  virtual void beginLocal() = 0;
+  virtual void beginExternal() = 0;
+  virtual bool connected() = 0;
+  virtual bool loop() = 0;
+  virtual bool connect(String hostname, String id, String password) = 0;
+  virtual void subscribe(String topic, int qos) = 0;
+  virtual void publish(String topic, String message) = 0;
+  virtual void onMessage(std::function<void(String, String)> callback) = 0;
+
+};
+
+//Wrapper for some of smartCar functionalities.
+struct SmartCarWrapper {
+
+    virtual ~SmartCarWrapper() = default;
+
+    virtual float getSpeed()           = 0;
+    virtual void setSpeed(float speed) = 0;
+    virtual void setAngle(int angle)   = 0;
+    virtual int getDistance()          = 0;
+    virtual void update()              = 0;
+};
+
+float convertSpeed(float speed);
+
+// In case of Arduino, use Arduino syntax,
+inline int stringToInt(String input) {
+#if defined(ARDUINO)
+  return input.toInt();
+#else // use the syntax for C++.
+  return std::stoi(input);
+#endif
+}
+
+// In this function is tested;
+// connect() to "arduino", "public", "public",
+// subscribe() to this "/smartcar/control/#" topic and subscribe() to this "/smartcar/connectionLost" topic.
+// and onMessage() function is tested to callback().
+// handleSpeedTopic() & handleAngleTopic() are commented out because they are tested in this method.
+void MQTTMessageInput(MqttWrapper &mqtt, SerialWrapper &serial){
```

Comment:
I am not familiar with testing in c++, but is it like in android or java where you have assertions?
Because I could not really see assertions in the tests case, so I am wondering how the test are considered passed or failed

Reply:
Yeah, they are similar. In this google unit test, we used, i.ex. EXPECT_EQ() this is a google function that expects two values in the parameters. These values should be equal. If the method parameter values which are given are equal then the test passed, otherwise in the case of inequality test fails.